### PR TITLE
Async combo field

### PR DESCRIPTION
**Pull request by TomTirapani**
Added AsyncComboField, a variation on ComboField which populates its items via a `queryAsync` function. The primary use case for this would be querying a remote server and returning a list of options.

Also added the option of overriding the default item renderer for fields which use it.

### CODE REVIEW

File: cmp/form/dropdown/AsyncComboField.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+ * @prop rest, see properties for HoistField
+ *
+ * @prop queryAsync, async function that receives (string query)
+ *          Should return a collection of form [{value: string, label: string}, ...] or [val, val, ...]
+ * @prop placeholder, text to display when control is empty
+ * @prop itemRenderer, optional custom itemRenderer, a function that receives (item, itemProps)
+ */
+@hoistComponent()
+export class AsyncComboField extends BaseDropdownField {
+    @observable.ref @setter items = [];
```
Comment (lbwexler) [inline cmp/form/dropdown/AsyncComboField.js]: Call this options, and normalize before setting, not at render time.

**Comment by TomTirapani** [comment]
Example usage:

```
comboField({
    model: model,
    field: 'value',
    queryAsync: (query) => {
        return XH.fetchJson({
            url: '...',
            params: {query: query}
        }).then(data => {
            return data;
        });
    }
});
```

File: cmp/form/dropdown/AsyncComboField.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+ *
+ * @prop rest, see properties for HoistField
+ *
+ * @prop queryAsync, async function that receives (string query)
+ *          Should return a collection of form [{value: string, label: string}, ...] or [val, val, ...]
+ * @prop placeholder, text to display when control is empty
+ * @prop itemRenderer, optional custom itemRenderer, a function that receives (item, itemProps)
+ */
+@hoistComponent()
+export class AsyncComboField extends BaseDropdownField {
```
Comment (lbwexler) [inline cmp/form/dropdown/AsyncComboField.js]: Not sure about the name of this component.    Its not just that its one time async loading, but it dynamically filters on every keystroke.

File: cmp/form/dropdown/AsyncComboField.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    static defaultProps = {
+        placeholder: 'Search'
+    }
+
+    delegateProps = ['className', 'style', 'placeholder', 'disabled'];
+
+    constructor(props) {
+        super(props);
+        this.addAutoRun(() => this.onQueryChange(), {delay: 100});
```
Comment (lbwexler) [inline cmp/form/dropdown/AsyncComboField.js]: should the delay be configurable?

File: cmp/form/dropdown/AsyncComboField.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        const {style, width, itemRenderer, disabled} = this.props;
+
+        const options = this.normalizeOptions(this.items),
+            value = this.renderValue;
+
+        return suggest({
+            popoverProps: {popoverClassName: Classes.MINIMAL},
+            $items: options,
+            onItemSelect: this.onItemSelect,
+            itemRenderer: itemRenderer || this.defaultItemRenderer,
```
Comment (lbwexler) [inline cmp/form/dropdown/AsyncComboField.js]: reading an overridden itemRenderer from props can probably be built into the shared getter

**Pull request merged**